File: Solutions/329. Longest Increasing Path in a Matrix/329. Longest Increasing Path in a Matrix.go
var dp [200][200]int
var res,m,n int

func longestIncreasingPath(mat [][]int) int {
    res,m,n=0,len(mat),len(mat[0])
    for i:=0;i<200;i+=1{
        for j:=0;j<200;j+=1{
            dp[i][j]=0
        }
    }
    for i:=0;i<m;i+=1{
        for j:=0;j<n;j+=1{
            res=max(res,dfs(&mat,i,j,-1))
        }
    }
    return res
}

func dfs(mat *[][]int,i,j,prev int) int{
    if i<0 || j<0 || i>=m || j>=n || (*mat)[i][j]<=prev{
        return 0
    }
    if dp[i][j]!=0{
        return dp[i][j]
    }
    dp[i][j]=1+max(max(dfs(mat,i+1,j,(*mat)[i][j]),dfs(mat,i-1,j,(*mat)[i][j])),max(dfs(mat,i,j+1,(*mat)[i][j]),dfs(mat,i,j-1,(*mat)[i][j])))
    return dp[i][j]
}

func max(i,j int) int{
    if i>j{
        return i
    }
    return j
}

File: Solutions/329. Longest Increasing Path in a Matrix/329. Longest Increasing Path in a Matrix.cpp
class Solution {
public:
    int dp[200][200]={};
    int res,m,n;
    int longestIncreasingPath(vector<vector<int>>& mat) {
        res=0,m=mat.size(),n=mat[0].size();
        for(int i=0;i<m;++i){
            for(int j=0;j<n;++j){
                res=max(res,dfs(mat,i,j,-1));
            }
        }
        return res;
    }
    int dfs(vector<vector<int>>& mat,int i,int j,int prev){
        if(i<0 or j<0 or i>=m or j>=n or mat[i][j]<=prev){
            return 0;
        }
        if(dp[i][j]){
            return dp[i][j];
        }
        return dp[i][j]=1+max({dfs(mat,i+1,j,mat[i][j]),
                              dfs(mat,i-1,j,mat[i][j]),
                              dfs(mat,i,j+1,mat[i][j]),
                              dfs(mat,i,j-1,mat[i][j])});
    }
};